Automattic/WP-Job-Manager
# Add support for activating multiple licenses at once

Fixes #2375
This depends on 221-gh-Automattic/wpjobmanager.com

### Changes proposed in this Pull Request

- Add support for bulk licensing multiple WPJM addons in a single request;
- Improve design of the Licenses page for WPJM addons;

### Testing instructions

Using a local version of WPJobManager.com running the latest code on 221-gh-Automattic/wpjobmanager.com, and this version of the WPJobManager plugin (and with the core addons enabled too):

1. Verify if you can activate a plugin;
2. Verify if you can bulk activate all the plugins;
3. Verify if the bulk activation form disappears when every plugin is disabled;
4. Verify what happens when there's an error connecting to your local WPJobManager.com server;
5. Verify if you can deactivate a plugin;
6. Verify what happens when you type a invalid code in the bulk activation form;
7. Verify what happens when you type a invalid code in the activation form for a plugin;
8. Verify if the design matches what was specified;
9. Try different variations of tests. it is very important to have this code well tested;

**Note:** You can use the following snippet to use a local WPJobManager.com version in your PC:

```php
add_filter( 'pre_http_request', function( $preempt, $r, $url ) {
	$host = parse_url( $url, PHP_URL_HOST );

	if ( 'wpjobmanager.com' === $host ) {
		$scheme  = parse_url( $url, PHP_URL_SCHEME );
		$http    = _wp_http_get_object();
		$new_url = str_replace(
			$scheme . '://' . $host,
			'http://YOUR_IP:YOUR_PORT',
			$url
		);

		error_log( 'Redirect from: ' . $url );
		error_log( 'To: ' . $new_url );

		return $http->request( $new_url, $r );
	}
	return $preempt;
}, 10, 3 );
add_filter( 'http_request_host_is_external', function($external) {
	return true;
});

add_filter( 'http_allowed_safe_ports', function($ports) {
	$ports[] = YOUR_PORT;
	return $ports;
});
```

### Notes

- I didn't add search on this PR nor added the separate view of inactive vs. active plugins. I'll add such tasks in smaller PRs, as this one is already quite big!
- ~The php unit checks here at github are failing because they're rely on an outdated mock. I'll checkout on fixing them later~ The tests are fixed;

### Screenshots

![image](https://user-images.githubusercontent.com/529864/231326808-44912ccc-f354-4d92-82a2-0300240617d5.png)

![image](https://user-images.githubusercontent.com/529864/231327260-750b39d1-1ac1-46fd-a818-69e1ed8653a1.png)
